project_2/include/simulator.hpp

```diff
@@ -0,0 +1,73 @@
+#pragma once
+
+#include "packet.hpp"
+#include "switch.hpp"
+#include "server.hpp"
+#include "link.hpp"
+
+#include <vector>
+#include <cstdint>
+#include <queue>
+#include <unordered_map>
+
+class NetworkSimulator {
+private:
+    static std::uint64_t id_to_give;
+
+    // Current time in simulation
+    std::uint64_t current_time_ns = 0;
+    // Total simulation duration
+    std::uint64_t simulation_duration_ns = 1000;
+
+    // Shows next time to call SendPacket method of hosts
+    // TODO: store this value for each host in some kind of priority_queue
+    std::uint64_t next_ask = 0;
+
+    // Storage of sending servers
+    std::unordered_map<std::uint64_t, std::shared_ptr<ServerBase>> ids_to_senders{};
+
+    // List of all links in network
+    std::vector<std::shared_ptr<Link>> links{};
+    
+    // Receiving server
+    std::shared_ptr<ServerBase> receiver;
+
+    // Switch in the middle between senders and receiver
+    std::shared_ptr<NetworkSwitch> n_switch;
+
+    // Sorts packages by sending time
+    std::priority_queue<RoutingPacket> packets{};
+    
+    /**
+     * Creates direct and reverse link between two network elements with given speed (in bits per ns)
+     *
+     * Do not forget to initialize corresponding routing tables after creating new connection
```

Comment:
Can we store corresponding routing table for `RoutingNetworkElement` in class and init them here?

Reply:
I am not sure, that I got your question correctly. Do you want to create routing table right here and then just pass it to RoutingNetworkElement class?

Reply:
My bad - thought that we can call something like:
```
src->AddNewRoute(...);
dest->AddNewRoute(...);
```
But as I see we sometimes need to add routes for all of sender, receiver and switch after adding new connection.